embassy-stm32/src/exti.rs

```diff
@@ -77,71 +71,136 @@ impl<T: gpio::ExtiPin + WithInterrupt + digital::InputPin> digital::InputPin for
     }
 }
 
-/*
-    Irq	Handler	Description
-    EXTI0_IRQn	EXTI0_IRQHandler	Handler for pins connected to line 0
-    EXTI1_IRQn	EXTI1_IRQHandler	Handler for pins connected to line 1
-    EXTI2_IRQn	EXTI2_IRQHandler	Handler for pins connected to line 2
-    EXTI3_IRQn	EXTI3_IRQHandler	Handler for pins connected to line 3
-    EXTI4_IRQn	EXTI4_IRQHandler	Handler for pins connected to line 4
-    EXTI9_5_IRQn	EXTI9_5_IRQHandler	Handler for pins connected to line 5 to 9
-    EXTI15_10_IRQn	EXTI15_10_IRQHandler	Handler for pins connected to line 10 to 15
-*/
-
-impl<T: gpio::ExtiPin + WithInterrupt + 'static> WaitForRisingEdge for ExtiPin<T> {
-    type Future<'a> = impl Future<Output = ()> + 'a;
-
-    fn wait_for_rising_edge<'a>(self: Pin<&'a mut Self>) -> Self::Future<'a> {
+impl<T: Instance + digital::InputPin + 'static> ExtiPin<T> {
+    fn wait_for_state<'a>(
+        self: Pin<&'a mut Self>,
+        state: EdgeOption,
+    ) -> impl Future<Output = ()> + 'a {
         let s = unsafe { self.get_unchecked_mut() };
 
-        s.pin.clear_interrupt_pending_bit();
+        s.pin.clear_pending_bit();
         async move {
             let fut = InterruptFuture::new(&mut s.interrupt);
             let pin = &mut s.pin;
             cortex_m::interrupt::free(|_| {
-                let mut exti: EXTI = unsafe { mem::transmute(()) };
-
-                pin.trigger_on_edge(&mut exti, Edge::RISING);
-                pin.enable_interrupt(&mut exti);
+                pin.trigger_edge(state);
             });
+
+            match state {
+                EdgeOption::Rising => {
+                    if s.pin.is_high().unwrap_or(false) {
+                        return;
+                    }
+                }
+                EdgeOption::Falling => {
+                    if s.pin.is_low().unwrap_or(false) {
+                        return;
+                    }
+                }
+                _ => unreachable!(),
```

Comment:
Why use `state: EdgeOption` + `unreachable` here and not `state: bool`?

Reply:
`EdgeOption` is already required for line 86. If I replaced it, I'd need another `match`.

Reply:
Also, `EdgeOption` is more descriptive than bool.

Reply:
`EdgeOption` isn't more descriptive since you're calling `wait_for_state` and passing in an "edge", not a "state"... it doesn't really make sense IMO, `EdgeOption` is "bigger" than type of states (`{high, low}` i.e. `bool`).

and this isn't `unreachable!` at all, I can just call `self.wait_for_state(EdgeOption::RisingFalling)`. it's a partial function.